ENH: Refactoring to support multiple Parquet readers, add PyArrow reader

wesm opened the pull request:
This is based on @mrocklin's initial patch #2175 and additional work from me (I needed to squash and rebase to resolve conflicts that happened in the meantime).

I have some changes in https://github.com/apache/arrow/pull/543 to help with this integration. The existing Fastparquet read path is unaffected (all the tests pass).

There's some notable things missing:

* Support writing the DataFrame index automatically, then reading it from the file (see below)
* I am not splitting files at the row group level; if this is needed, it can be added in a follow up patch
* Categorical support (Parquet doesn't have a Categorical type, but there's workarounds to simulate it that are in use in Dask)

There's a couple issues we ought to resolve to conform the semantics of how pandas.DataFrame is written to Parquet, and how we read it:

* Index handling, inferring the index column, and other metadata: #2222 
* The `has_nulls` argument -- this doesn't impact the pyarrow read path, but if we want to support pyarrow writes, then it would be good to do the same thing by default across implementations: #2177 

On the pandas index, I think the most reliable thing we could do is to write special metadata to the Parquet file footer to indicate that a column is the index. This is a lot cleaner than inferring based on the column name or the column statistics. 

My goal in subsequent patches will be to add the `engine` parameter to the test cases where it isn't being used. Thanks!

mrocklin commented:
Thanks for putting this together @wesm .  I'm looking forward to trying this out tomorrow.

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +249,171 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+        def normalize_PyArrowParquetDataset(ds):
+            return (type(ds), ds.paths)
+
+        self._REGISTERED_TOKENIZE = True
```
Is there a reason to put this within the class method instead of at top-level?

As is this code only gets called if someone calls `read_parquet` and not if they're using dask.delayed or some other dask structure with `pyarrow.ParquetDataset` objects.

One possible objective would be to hide the import cost until necessary, but in this case it doesn't seem to be that bad

```python
In [1]: import numpy, pandas

In [2]: %time import pyarrow.parquet
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 19.3 ms
```

wesm replied:
I put these back at the end of the file

mrocklin commented:
We'll also want to think about column access optimizations.  Currently dask.dataframe does a *tiny* bit of optimization to convert the following:

```python
df = dd.read_parquet(fn)
s = df.my_column
```

to

```python
s = dd.read_parquet(fn, columns='my_column')
```

See https://github.com/dask/dask/blob/master/dask/dataframe/optimize.py#L23-L24

mrocklin commented:
@martindurant you will probably want to take a look at this when you have some time.

mrocklin commented:
Also @wesm if you do produce a dev build somewhere I would enjoy taking a deeper look.  I'm playing with a large parquet dataset now for other purposes (some machine learning work) and would happily be a guinea pig if convenient.

wesm commented:
> Style-wise I would prefer to see these as functions, though this is subjective.

I generally prefer functions, too, but not long functions that do a long list of things. 

If you look at the existing `read_parquet` function in the codebase, here is a list of things it does, all in a single 100 line function:

* Checks if fastparquet is installed
* User argument validation
* Resolves the file system and paths based on the storage options
* Computes the Dask task name
* Filters out row groups based on predicates
* Infer the index column based on column names or statistics
* Decides which columns to read from the file based on user input
* Constructs dummy metadata for the Dask DataFrame constructor
* Addresses Series vs DataFrame return type
* constructs the Dask task manifest
* Computes divisions fro the resulting Dask dataframe

That's 10 things in a single function; this pattern makes refactoring difficult and readability difficult for outsiders (I rearranged the code in part so I could understand everything that it's doing).

You can split this into a bunch of functions, but sometimes you fall into the "all my functions have 10 arguments" trap that plagues functional programmers. I can make everything functions if it's preferred, but I argue that what is in this patch is strictly better than one monolithic function.

wesm commented:
From Chapter 3 of *Clean Code* by Uncle Bob Martin, one of my favorite books on software engineering

> It should be very clear that Listing 3-1 is doing lots more than one thing. It’s creating buffers, fetching
> pages, searching for inherited pages, rendering paths, appending arcane strings, and generating HTML,
> among other things. Listing 3-1 is very busy doing lots of different things. On the other hand, Listing 3-3
> is doing one simple thing. It’s including setups and teardowns into test pages. The following advice has appeared in one form or another for 30 years or more.
>
> FUNCTIONS SHOULD DO ONE THING . THEY SHOULD DO IT WELL .
> THEY SHOULD DO IT ONLY .

mrocklin commented:
With due respect to Bob Martin there are also costs to indirection.  When maintaining code I find it challenging to constantly navigate to different parts of the file to follow what is actually a single linear flow.  If functions will be used exactly once then I would rather see them inlined, perhaps with a single comment line giving high-level understanding.

Again though, I think that this is subjective.  Different people have different styles and no style is strictly superior to another.  If you're planning to maintain this code long term then I don't mind if this follows a different style.  If I'm expected to maintain this code long term then I would prefer to avoid indirection.  If, as we both hope, a general community is likely to maintain the code then there should probably be a broader conversation about style.  Until that happens I'd prefer if we kept to conventions common within the codebase today.

mrocklin commented:
@martindurant also maintains this code a bit, I'd be happy to yield if he prefers things broken up a bit as is done here.

wesm commented:
It is a subjective view, and I think inlining works up until a point. Eventually, unless you are building a one-off tool that doesn't get developed that much, I have often found you hit a maintainability breaking point and have to refactor. I suspect that this code started out with fewer components and has progressively had more features tacked on. `pandas.read_csv` is one example of a function that started out as one inlined thing and eventually had to be modularized for the sake of maintainability (and now it's a 3K line behemoth with lots of functions encapsulating low level details https://github.com/pandas-dev/pandas/blob/master/pandas/io/parsers.py). 

I think the cognitive load of long functions disproportionately affects the non-primary authors. So if you wrote the code, you can navigate things more easily than others. I've always been a proponent of anything that improves code readability for other people: descriptive variable names, descriptive function names, modularization, code reuse. Long functions and non-descriptive variable names harkens back to MATLAB where anything else was penalized by the programming environment. 

I do plan to help maintain this code; if you all want one large function, I'm willing to conform to that style with my objections noted...

mrocklin commented:
> It is a subjective view, and I think inlining works up until a point. Eventually, unless you are building a one-off tool that doesn't get developed that much

This is a one-off tool though.  The goal of this function is to take a path and produce a dask.dataframe.  I agree that there would be reason to split things up if we were going to use the various break-out points that multiple functions or a class of many methods would provide.

> I have often found you hit a maintainability breaking point and have to refactor.

I agree.  I personally tend to switch to breaking things down when I find I need to test them separately in order to develop effectively.  I don't think that we're there yet with this (although I could be wrong).

> I think the cognitive load of long functions disproportionately affects the non-primary authors. So if you wrote the code, you can navigate things more easily than others. 

Interesting, I feel the exact same way about hierarchical structures.  Probably both are true.  Generally I'm in favor of using simple tools when they suffice, rather than constructing APIs and systems that have to be learned before code can be understood.  I generally try to invent as little as necessary.

> I've always been a proponent of anything that improves code readability for other people: descriptive variable names, descriptive function names, modularization, code reuse.

Yes, my motivations in this discussion are the exact same as yours here.

> Long functions and non-descriptive variable names harkens back to MATLAB where anything else was penalized by the programming environment

I'm sorry if you thought I was encouraging non-descriptive variable names.  That was not my intent.  

I'm also not encouraging long functions as virtuous.  I agree with you that they come at a cost.  I just find indirection and building up class hierarchies to be a higher cost.  Generally I don't think that functions and classes should be invoked strictly for documentation purposes.  I think that this benefit (which I agree exists) doesn't outweigh the cost of having to navigate a linear problem with non-linear code.  

We've obviously both programmed a bit and had different experiences.  I think that both experiences are valid.  Probably we can come to some better solution through cooperation than either of us would have produced independently.  

Again, if you're planning to maintain most of this then I'm happy with things the way they are now. 
 @martindurant probably spends more maintenance time than I do, and I suspect he'll probably end up maintaining the fastparquet portion more than either of us (unless you're also planning to take this on).  So again I'm quite happy to go with the current design if you all prefer.

martindurant commented:
I am of course very biased here, so my opinion should be taken with salt. Where the line is between something I can understand, fit on the screen, and quickly change, and something other contributors can quickly comprehend is hazy. That haze is redoubled when the potential contributors themselves have different styles.

Generally speaking, I agree with @mrocklin , which should not be a surprise from the code I've written. I see no benefit in a class for class's sake, and have very firm ideas about what a class should be for (no need to go into that here). There are many possible measures for complexity of code from number of lines and variables to depth of indents and such... 

In this particular case, I would stick with the functions pattern mainly because it fits in with the style elsewhere in dask (and fastparquet). The many-variables and many-lines nature of the functions are a hindrance, but I don't feel that the classes are semantically simpler. An intermediate state would be splitting long functions into shorter ones, and passing **kwargs or a configure dict around, but, again, I see no benefit of `def func(abc)` over `# now use abc` *unless* we can make tests for the components.

wesm commented:
No problem. I reverted things to the way they were with a couple small overlapping bits factored out. 

I also added an xfailing unit test for #2177, so figure we address that after this goes in. 

```
@pytest.mark.skipif(not pyarrow, reason='pyarrow not found')
@pytest.mark.xfail(reason="to_parquet does not write nulls by default")
def test_to_parquet_default_writes_nulls():
    import pyarrow.parquet as pq

    df = pd.DataFrame({'c1': [1., np.nan, 2, np.nan, 3]})
    ddf = dd.from_pandas(df, npartitions=1)

    with tmpfile() as fn:
        ddf.to_parquet(fn)
        table = pq.read_table(fn)
        assert table[0].null_count == 2
```

wesm commented:
Also, I can add pyarrow to Dask's CI after the conda-forge packages get updated to the latest

wesm commented:
The Appveyor failure seems unrelated. Rebasing in case that fixes it

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +147,147 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+# User read API
+
+def read_parquet(path, columns=None, filters=None, categories=None, index=None,
+                 storage_options=None, engine='fastparquet'):
```
We could change `engine=` to `'auto'` that would use whichever library was present.

wesm replied:
done

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +147,147 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+    if isinstance(columns, tuple):
+        columns = list(columns)
+
+    columns = columns
```
This can probably be removed

wesm replied:
done

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +147,147 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+    schema = dataset.schema.to_arrow_schema()
+    task_name = 'read-parquet-' + tokenize(dataset, columns)
+
+    pieces = dataset.pieces
```
Style nit: Maybe just use `dataset.pieces` below and strike this line?

wesm replied:
done

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +147,147 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+    else:
+        out_type = DataFrame
+
+    divisions = (None,) * (len(dataset.pieces) + 1)
```
We're waiting on the Pandas metadata discussion for auto-indexing I take it?

wesm replied:
Yep. We should really hustle to do this ASAP so we can do the work in both fastparquet and pyarrow to implement the standard index metadata

mrocklin commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +147,147 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+                         file_opener,
+                         piece, all_columns,
+                         out_type == Series,
+                         dataset.partitions)
```
What is `dataset.partitions` here?  Also what is piece?  Mostly I ask to ensure that they are both individually cheap to serialize.  Every task gets pickled (or cloudpickled) independently.  We've had issues in the past from putting full-dataset metadata into every task which then caused serialization time to become annoyingly large.

wesm replied:
It's a small data structure containing the partition set; its size is proportional to the number of partitions. I haven't optimized its serialization time, but if it seems to be affecting things with large # of partitions:

```python
In [14]: p1 = pq.PartitionSet('foo', keys=['foo_{0}' for i in range(10000)])

In [15]: p2 = pq.PartitionSet('bar', keys=['bar{0}' for i in range(1000)])

In [16]: partitions = pq.ParquetPartitions()

In [17]: partitions.levels = [p1, p2]

In [18]: timeit pickle.dumps(partitions)
10000 loops, best of 3: 152 µs per loop
```

wesm replied:
piece is even simpler, it is: file path, row group number, and partition keys for the file. Basically these are both tools to have to pass around fewer arguments so I don't have to thread parameters through functions to pass around additional metadata

FWIW: the current fastparquet task executor might benefit from a helper data structure to bundle up some of the parameters:

```
def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
                            schema, cs, dt, *args):
    ...
```

mrocklin replied:
OK, it looks like serializing that a few thousand times will probably not be a dominant cost.  Regardless this is probably premature optimization.

Have you been trying dask.dataframe + pyarrow on data outside of tests?  If so any experience to report?  You might want to try engaging the distributed scheduler as well just on your local laptop to simulate the more complex environment

```python
from dask.distributed import Client
client = Client()  # Starts local "cluster".  Registers itself as default global scheduler

# do normal dask.dataframe.read_parquet call as before
```

wesm replied:
I'll give this a try when I can

mrocklin commented:
Everything here looks good to me.  I think we're currently waiting both for @martindurant to have a chance to look things over and also for compatible pyarrow builds on conda-forge.

wesm commented:
If you'd like to try things out yourself, you can build things on Linux and OS X (and set up your toolchain for subsequent builds) by following this guide: https://github.com/apache/arrow/pull/548. We'll get up `0.3.pre` builds this week so we can add to CI

martindurant commented:
I have no further criticisms of the code. Do we need to put something in the docs (as opposed to just docstrings) ?

wesm commented:
conda-forge packages are building right now (https://github.com/conda-forge/pyarrow-feedstock/commit/b9844e4ceae850c2fff1c32350c9da437d9f4600), I'll add pyarrow to Dask's CI in the next couple days

wesm commented:
Just rebased and added pyarrow to the Travis install script

wesm commented:
Next steps from here:

* Specify KeyValue metadata for the index column(s), push down this logic into the reader libraries when `index=None`
* Specify metadata for pandas-specific types, like categorical

Any other metadata we want to codify / standardize? I can start a PR to address these per https://github.com/pandas-dev/pandas/issues/16010 and we can discuss there?

martindurant commented:
On categoricals, the following information could be considered:
- which columns are intended to be interpreted with categorical dtype (obviously they can use dictionary encoding regardless)
- the dtype of the column to store the cat codes (1, 2, 4, 8-byte ints), which in fastparquet we are now storing as the number of expected labels per column
- whether each partition of the data has exactly the same categories, which they will if they are derived from a single original data-frame; if not, the category counts for each partition. This might be stored in the global metadata, in the column chunk metadata, or in the individual files. I weakly prefer the first option.

On the index, I suggest *not* using the parquet sorting column field, since sorting and (partition-wise) indexing are not quite the same thing.

wesm commented:
My thought with categorical was to put key-value metadata like

```
{
  'type': 'categorical',
  'index_type': 'int16',
  'num_categories': K
}
```

for the index, we can put this info in the file-level key-value metadata https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L620 as

```
{
  'index_columns': [
    {'column': '__index_level_0__',
     'name': 'index'}
  ]
}
```

The particulars don't matter too much as long as we encode complete and unambiguous metadata.

wesm commented:
> whether each partition of the data has exactly the same categories, which they will if they are derived from a single original data-frame; if not, the category counts for each partition. 

How would this information be used? Keeping track of the partition category metadata in the global metadata seems potentially brittle; in theory the code users are writing should be robust to this (if categoricals are being used effectively as a compression technique)

mrocklin commented:
I tried reading some parquet data locally with both fastparquet and arrow.  I found that arrow was much faster to read all of the metadata (like 5x-10x faster), but that it seg faulted when I tried to get data out:

```python
>>> import dask.dataframe as dd
df = dd.read>>> df = dd.read_parquet('day-0-bytes.parquet', engine='arrow')
>>> df.head()
[New Thread 0x7fffdf6cf700 (LWP 9073)]
[New Thread 0x7fffdeece700 (LWP 9074)]
[New Thread 0x7fffde6cd700 (LWP 9075)]
[New Thread 0x7fffddecc700 (LWP 9076)]
[New Thread 0x7fffdd6cb700 (LWP 9077)]
[New Thread 0x7fffdceca700 (LWP 9078)]
[New Thread 0x7fffc7fff700 (LWP 9079)]

Thread 2 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffdf6cf700 (LWP 9073)]
parquet::TypedColumnReader<parquet::DataType<(parquet::Type::type)6> >::ReadBatch (this=0x0, 
    batch_size=batch_size@entry=253545, def_levels=def_levels@entry=0x7fffdb820b80, 
    rep_levels=rep_levels@entry=0x0, values=values@entry=0x7fffdc2ab040, 
    values_read=values_read@entry=0x7fffdf6ccd90)
    at /feedstock_root/build_artefacts/parquet-cpp_1492652141393/work/parquet-cpp-5830482f7cc11392f9a14d9520bcbd8d6adc944b/src/parquet/column/reader.h:223
223	/feedstock_root/build_artefacts/parquet-cpp_1492652141393/work/parquet-cpp-5830482f7cc11392f9a14d9520bcbd8d6adc944b/src/parquet/column/reader.h: No such file or directory.
```

I had conda installed pyarrow from conda-forge.  Perhaps I have an old version of parquet hanging around somewhere?

mrocklin commented:
Some information about the file:

<details>

```python
In [1]: import fastparquet

In [2]: pf = fastparquet.ParquetFile('day-0-bytes.parquet/')

In [3]: pf
Out[3]: <Parquet File: {'name': 'day-0-bytes.parquet//_metadata', 'columns': ['click', 'numeric_0', 'numeric_1', 'numeric_2', 'numeric_3', 'numeric_4', 'numeric_5', 'numeric_6', 'numeric_7', 'numeric_8', 'numeric_9', 'numeric_10', 'numeric_11', 'numeric_12', 'category_0', 'category_1', 'category_2', 'category_3', 'category_4', 'category_5', 'category_6', 'category_7', 'category_8', 'category_9', 'category_10', 'category_11', 'category_12', 'category_13', 'category_14', 'category_15', 'category_16', 'category_17', 'category_18', 'category_19', 'category_20', 'category_21', 'category_22', 'category_23', 'category_24', 'category_25'], 'partitions': [], 'rows': 195841983}>

In [4]: pf.dtypes
Out[4]: 
{'category_0': dtype('O'),
 'category_1': dtype('O'),
 'category_10': dtype('O'),
 'category_11': dtype('O'),
 'category_12': dtype('O'),
 'category_13': dtype('O'),
 'category_14': dtype('O'),
 'category_15': dtype('O'),
 'category_16': dtype('O'),
 'category_17': dtype('O'),
 'category_18': dtype('O'),
 'category_19': dtype('O'),
 'category_2': dtype('O'),
 'category_20': dtype('O'),
 'category_21': dtype('O'),
 'category_22': dtype('O'),
 'category_23': dtype('O'),
 'category_24': dtype('O'),
 'category_25': dtype('O'),
 'category_3': dtype('O'),
 'category_4': dtype('O'),
 'category_5': dtype('O'),
 'category_6': dtype('O'),
 'category_7': dtype('O'),
 'category_8': dtype('O'),
 'category_9': dtype('O'),
 'click': dtype('int64'),
 'numeric_0': dtype('float64'),
 'numeric_1': dtype('float64'),
 'numeric_10': dtype('float64'),
 'numeric_11': dtype('float64'),
 'numeric_12': dtype('float64'),
 'numeric_2': dtype('float64'),
 'numeric_3': dtype('float64'),
 'numeric_4': dtype('float64'),
 'numeric_5': dtype('float64'),
 'numeric_6': dtype('float64'),
 'numeric_7': dtype('int64'),
 'numeric_8': dtype('int64'),
 'numeric_9': dtype('float64')}

In [5]: pf.row_groups[0]
Out[5]: 
<class 'parquet_thrift.RowGroup'>
columns: [<class 'parquet_thrift.ColumnChunk'>
file_offset: 2028397
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 4
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['click']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: 
    min: 
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 2

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 4056866
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 2028473
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_0']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 6085342
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 4056949
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_1']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 8113818
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 6085425
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_2']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 10142294
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 8113901
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_3']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 12170770
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 10142377
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_4']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 14199246
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 12170853
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_5']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 16227722
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 14199329
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_6']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 18256198
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 16227805
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_7']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: L
    min: b'\xff\xff\xff\xff\xff\xff\xff\xff'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 2

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 20284674
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 18256281
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_8']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: A
    min: 
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 2

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 22313150
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 20284757
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_9']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 24341626
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 22313233
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_10']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 26370103
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 24341710
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_11']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 28398580
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 26370187
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['numeric_12']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    min: b'\x00\x00\x00\x00\x00\x00\xf8\x7f'
    null_count: 0

  total_compressed_size: 2028393
  total_uncompressed_size: 2028393
  type: 5

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 30396118
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 28398664
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_0']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 32424584
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 30396183
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_1']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ffe0ddaa
    min: 00007246
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 34453069
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 32424668
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_2']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fff0eed3
    min: 0000803f
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 36481554
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 34453153
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_3']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ffe2b3ca
    min: 001fe4bd
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 38510039
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 36481638
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_4']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ffd81782
    min: 0001c733
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 40538524
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 38510123
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_5']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: 919c68e4
    min: 6fcd6dcb
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 42567009
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 40538608
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_6']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fff605eb
    min: 000112e6
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 44595494
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 42567093
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_7']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ffac3596
    min: 00900e80
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 46623979
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 44595578
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_8']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ff6bef7c
    min: 0f598795
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 48621517
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 46624063
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_9']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 50619036
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 48621582
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_10']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 52647503
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 50619102
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_11']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fffcee4d
    min: 00042473
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 54675989
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 52647588
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_12']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: f4fc3c5e
    min: 00c5ffb7
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 55931424
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 54676074
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_13']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 100593

  total_compressed_size: 1255350
  total_uncompressed_size: 1255350
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 57959892
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 55931491
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_14']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fff9fcb3
    min: 002b0cc2
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 59215327
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 57959977
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_15']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 100593

  total_compressed_size: 1255350
  total_uncompressed_size: 1255350
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 60470744
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 59215394
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_16']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 100593

  total_compressed_size: 1255350
  total_uncompressed_size: 1255350
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 62499212
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 60470811
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_17']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ffecbda7
    min: 024b6f0d
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 64527698
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 62499297
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_18']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ff4b7387
    min: 09c8c22c
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 66525237
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 64527783
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_19']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 68522757
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 66525303
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_20']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 70520277
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 68522823
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_21']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 7830

  total_compressed_size: 1997454
  total_uncompressed_size: 1997454
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 71775693
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 70520343
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_22']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: None
    min: None
    null_count: 100593

  total_compressed_size: 1255350
  total_uncompressed_size: 1255350
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 73804161
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 71775760
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_23']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fffe3770
    min: 00140db7
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 75832647
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 73804246
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_24']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: ff654802
    min: 0a95e0db
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

, <class 'parquet_thrift.ColumnChunk'>
file_offset: 77861133
file_path: part.0.parquet
meta_data: <class 'parquet_thrift.ColumnMetaData'>
  codec: 0
  data_page_offset: 75832732
  dictionary_page_offset: None
  encoding_stats: [<class 'parquet_thrift.PageEncodingStats'>
count: 1
encoding: 0
page_type: 0
]
  encodings: [3, 4, 0]
  index_page_offset: None
  key_value_metadata: []
  num_values: 253545
  path_in_schema: ['category_25']
  statistics: <class 'parquet_thrift.Statistics'>
    distinct_count: None
    max: fdfc9d62
    min: 0b919f8e
    null_count: 0

  total_compressed_size: 2028401
  total_uncompressed_size: 2028401
  type: 7

]
num_rows: 253545
sorting_columns: None
total_byte_size: 77858042
```

</details>

Also pyarrow fails at reading it in as a table

```python
In [3]: table = pq.read_table('day-0-bytes.parquet/')
Segmentation fault (core dumped)
```

mrocklin commented on continuous_integration/travis/install.sh:
```diff
@@ -62,7 +62,7 @@ pip install -q git+https://github.com/dask/zict --upgrade --no-deps
 
 if [[ $PYTHONOPTIMIZE != '2' ]]; then
-    conda install -q -c conda-forge numba cython
+    conda install -q -c conda-forge numba cython pyarrow
```
Why did you put pyarrow here instead of above?

wesm replied:
The tests won't run if fastparquet is not installed, I'll move this to above

mrocklin commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -106,12 +121,18 @@ def test_series(fn):
-    assert set(read_parquet(fn).dask) == set(read_parquet(fn).dask)
+def test_names(fn, engine):
+    def read(fn):
+        return read_parquet(fn, engine=engine)
```
Did you mean to use `read` below here rather than `read_parquet`?

wesm replied:
Yep, fixed

mrocklin commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -369,3 +390,17 @@ def test_timestamp_index():
+
+
+@pytest.mark.skipif(not pyarrow, reason='pyarrow not found')
+@pytest.mark.xfail(reason="to_parquet does not write nulls by default")
```
Given recent changes do we still need to xfail this test?

martindurant replied:
The default changed in fastparquet, haven't looked at the call in dask yet (since it's been changing).

wesm replied:
Perhaps we can remove this xfail when making the change to the argument in Dask as proof that it's doing the expected thing.

wesm commented:
@mrocklin is there some way for me to reproduce the segfault? I can dig into it and see what's going on

wesm commented:
The parquet-cpp version should be 1.1.0.pre, I thought I had this pinned correctly in the meta.yaml, but maybe not.

mrocklin commented:
> @mrocklin is there some way for me to reproduce the segfault? I can dig into it and see what's going on

If you want to download and play with data I'm working from the criteo ML dataset (fairly large).  

    wget http://azuremlsampleexperiments.blob.core.windows.net/criteo/day_1.gz
    gunzip day_1.gz

```python
import dask.dataframe as dd
from dask.distributed import Client
client = Client()

categories = ['category_%d' % i for i in range(26)]
columns = ['click'] + ['numeric_%d' % i for i in range(13)] + categories

df = dd.read_csv('day_0', sep='\t', names=columns, header=None) 

encoding = {c: 'bytes' for c in categories}
fixed = {c: 8 for c in categories}
df.to_parquet('day-0-bytes.parquet', object_encoding=encoding,
              fixed_text=fixed, compression='SNAPPY')
```

The file is 24 GB, 47 GB uncompressed.  The conversion takes tens of minutes.  Alternatively if I have some time tomorrow I'll try reducing it down.

martindurant commented:
`'bytes'`: did you not have joy with categories?

mrocklin commented:
I did not have joy with categories.  Although things may have improved since then.

wesm commented:
Cool, I will download this dataset later at home on faster internet and give it a shot. Would be good to have some datasets like this around to use for benchmarking / optimization (e.g. we're going to add a strings-as-categorical option to the reader in the reasonably near future, this seems like a good test case)

mrocklin commented:
Yeah, I've found it a very valuable dataset.  It has been driving some of
my development recently.  It also grows to 1TB if desired.

The text columns have very different entropies.  Some have just a few
categories while some have millions.

On Tue, Apr 25, 2017 at 3:47 PM, Wes McKinney <notifications@github.com>
wrote:

> Cool, I will download this dataset later at home on faster internet and
> give it a shot. Would be good to have some datasets like this around to use
> for benchmarking / optimization (e.g. we're going to add a
> strings-as-categorical option to the reader in the reasonably near future,
> this seems like a good test case)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/2223#issuecomment-297144179>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszAWd11YFXNsmka4TDtn71eruYb4Nks5rzk3jgaJpZM4M-Oqe>
> .
>

wesm commented:
I tried running the code from both dask and distributed master and got

```
/home/wesm/anaconda3/envs/py27/lib/python2.7/site-packages/tornado/gen.pyc in wrapper(*args, **kwargs)
    305                 try:
    306                     orig_stack_contexts = stack_context._state.contexts
--> 307                     yielded = next(result)
    308                     if stack_context._state.contexts is not orig_stack_contexts:
    309                         yielded = TracebackFuture()

/home/wesm/code/dask-distributed/distributed/deploy/local.pyc in _start(self, ip)
    137         if self.status == 'running':
    138             return
--> 139         if ip is None and not self.scheduler_port and not self.processes:
    140             # Use inproc transport for optimization
    141             scheduler_address = 'inproc://'

AttributeError: 'LocalCluster' object has no attribute 'scheduler_port'
```

Not sure what I did wrong, let me know

wesm commented:
Seems it was a problem with the pip requirements, I did `conda install dask` (then removed dask/distributed manually) and I think I'm in business

mrocklin commented:
You might want to navigate to http://localhost:8787/status to get feedback

On Tue, Apr 25, 2017 at 10:07 PM, Wes McKinney <notifications@github.com>
wrote:

> Seems it was a problem with the pip requirements, I did conda install dask
> (then removed dask/distributed manually) and I think I'm in business
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/2223#issuecomment-297216880>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszDUjNH2Q4wZYQZMbXSFSc_IemCPSks5rzqbygaJpZM4M-Oqe>
> .
>

wesm commented:
I'm able to reproduce the segfault, will investigate.

As one side note: for local disk, it's less efficient to read using Python file objects in the pyarrow reader than to use its native file IO (which uses mmap to handle data buffering). I don't know if there's machinery in place to handle these kinds of operations.

wesm commented:
@mrocklin I found the problem -- fastparquet is writing text data with `FIXED_LEN_BYTE_ARRAY` type. While this is not technically incorrect, it seems atypical. I am not sure these files can be read consistently by other Parquet consumers. Should definitely support reading this data, though (and not segfault) -- I will take care of it in as soon as I can:

https://issues.apache.org/jira/browse/PARQUET-965

mrocklin commented:
I think that @martindurant found that fixed-length bytestrings were much
faster in the Python/NumPy code than were variable length bytestrings.
Given how numpy handles fixed length vs variable length bytes objects I
don't find this surprising.  I can also try avoiding fixed-length and see
what the performance hit is.  Honestly for my current application parquet
isn't really the bottleneck.

On Tue, Apr 25, 2017 at 10:35 PM, Wes McKinney <notifications@github.com>
wrote:

> @mrocklin <https://github.com/mrocklin> I found the problem --
> fastparquet is writing text data with FIXED_LEN_BYTE_ARRAY type. While
> this is not technically incorrect, it seems atypical. I am not sure these
> files can be read consistently by other Parquet consumers. Should
> definitely support reading this data, though (and not segfault) -- I will
> take care of it in as soon as I can:
>
> https://issues.apache.org/jira/browse/PARQUET-965
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/2223#issuecomment-297220569>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszH7CAX-Gs8H80WpJautLq_USEdnZks5rzq1ggaJpZM4M-Oqe>
> .
>

martindurant commented:
>> whether each partition of the data has exactly the same categories, which they will if they are  derived from a single original data-frame; if not, the category counts for each partition.

> How would this information be used? Keeping track of the partition category metadata in the global metadata seems potentially brittle; in theory the code users are writing should be robust to this (if categoricals are being used effectively as a compression technique)

The problem here, is that when writing parquet from dask, the dask-dataframe metadata may know that a column is supposed to be categorical, but will not generally have the labels available, and those labels may be different in each partition. When loading back each partition, we would like to know the appropriate dtype to assign for the category codes (per partition) before loading anything, and ideally don't want to have to probe the metadata of each constituent file to do so (because it adds overhead, potentially substantial on S3). 
This problem does not exist for writing from a pandas dataframe, where we know a priori that the category labels are the same everywhere.

None of this affects the case where dictionary encoding is used as compression for (some of) a column which is not intended to be interpreted as categorical - there is no need to store extra metadata for that. 
For the above case of whether to use fixed-length bytestrings, categorical encoding will be the best option at least in the case where the cardinality isn't excessive, but cannot be done until this metadata issue is resolved.

mrocklin commented:
To be clear I don't think that this PR needs to block on anything like fixed length bytestring support in arrow.  There are some more mundane tests failing though that, I think, have fairly straightforward solutions.

@martindurant from your perspective as a Dask maintainer is there anything else that you would like to see here?

mrocklin commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -106,14 +121,20 @@ def test_series(fn):
-            set(read_parquet(fn, columns=['x']).dask))
+def test_names(fn, engine):
+    def read(fn):
+        return read_parquet(fn, engine=engine)
```
Looks like this needs to pass through kwargs.  Or perhaps just define the following:

    read_parquet = partial(dd.read_parquet, engine=engine)

martindurant replied:
Agree with @mrocklin ; plus suggest the test could have a better name. We are really testing the passthrough of getitem, right?

martindurant commented on dask/dataframe/io/parquet.py:
```diff
@@ -174,11 +145,155 @@ def _read_parquet_row_group(open, fn, index, columns, rg, series, categories,
+    storage_options : dict
+        Key/value pairs to be passed on to the file-system backend, if any.
+    engine : {'auto', 'fastparquet', 'arrow'}, default 'auto'
+        Parquet reader library to use. If only one library is installed, it
```
This phrasing seems wrong. Presumably if I say "arrow" but it isn't around, there is an error, and the If Only One relates to the 'auto' choice.

martindurant commented on dask/dataframe/io/tests/test_parquet.py:
```diff
@@ -106,14 +121,20 @@ def test_series(fn):
+    assert (set(read(fn).dask) !=
+            set(read(fn, columns=['x']).dask))
+
+    assert (set(read(fn, columns='x').dask) !=
```
I'm surprised if this works - probably because the string is iterable (i.e., would fail if the column name was not one-character long).

wesm replied:
Removed this test ccase

wesm commented:
Thanks. I'll update the patch later today to make the build pass, and set has_nulls=True. I can also add a note to the docs

wesm commented:
I added some documentation and fixed the unit test. Thanks!

mrocklin commented:
+1 from me.  @martindurant want to give this a final look and merge if you're everything looks OK to you?

martindurant commented:
lgtm

martindurant commented:
(@mrocklin, didn't press merge, because I'm not sure if you'd want to squash or not)

mrocklin commented:
Generally we've been squashing by default for a while now.  Merging.  Thanks @wesm !

wesm commented:
thanks! Now we should address the index metadata discussion, I will follow up in that issue.
